### PR TITLE
[NPUW] DCOFF + Transpose

### PR DESCRIPTION
**Pull request by smirnov-alexey**

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp [hunk 1]
```diff
@@ -362,6 +323,36 @@ std::size_t LazyTensorImpl::get_hash() const {
...
 
+void LazyTensorImpl::get_transformations(std::vector<LazyTensor::Transform>& vec) const {
+    vec.push_back(m_transform);
+    std::visit(overloaded{[&vec](const op::Concat& op) {
+                              for (const auto& lt : op.tensors) {
+                                  auto next_tr = lt.get_transformations();
+                                  vec.insert(vec.end(), next_tr.begin(), next_tr.end());
+                              }
+                          },
+                          [&vec](const op::Const& op) {
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:334]: vec not used warning

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp [hunk 2]
```diff
@@ -139,15 +135,21 @@ void apply_remap(Subgraph& fcall, const ClosureRemap& m) {
         if (fcall._closure[i]) {
             new_closure.push_back(fcall._closure[i]);
         } else {
-            new_closure.push_back(m.weights_to_unpack.count(i) ? fcall._lazy_closure[i].eval() : fcall._closure[i]);
+            // Only support permute transformations
+            auto transforms = fcall._lazy_closure[i].get_transformations();
+            if (transforms.size() == 2 && std::holds_alternative<ov::npuw::weights::op::Permute>(transforms[0])) {
+                // Assuming we only match weights where DQ did set Transpose after - thus nothing extra to do here
+                new_closure.push_back(ov::Tensor());
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp:142]: Since it's going to be reused via DCOFF, we should mark those somehow so those will be allocated on CPU

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp:142]: But in this case they won't be actually shared in the bank...

File: src/frontends/common/src/input_model.cpp [hunk 3]
```diff
@@ -12,7 +12,9 @@
 using namespace ov;
 using namespace ov::frontend;
 
-InputModel::~InputModel() = default;
+InputModel::~InputModel() {
+    std::cout << "Input model destructor" << std::endl;
```
Comment (praasz) [inline src/frontends/common/src/input_model.cpp:16]: Remove it

**Comment by smirnov-alexey** [comment]
@praasz I'm currently debugging memory consumption. Sorry for those debug prints, I'll remove them once done

**Comment by praasz** [comment]
> @praasz I'm currently debugging memory consumption. Sorry for those debug prints, I'll remove them once done

Consider add labels `WIP`, `do_not_review` if PR is still in development progress.

File: src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp [hunk 4]
```diff
@@ -139,17 +135,34 @@ void apply_remap(Subgraph& fcall, const ClosureRemap& m) {
...
+            auto transforms = fcall._lazy_closure[i].get_transformations();
+            if (transforms.size() == 2 && std::holds_alternative<ov::npuw::weights::op::Permute>(transforms[0])) {
+                // Assuming we only match weights where DQ did set Transpose after - thus nothing extra to do here
+                std::cout << "NOT EVALUATING TENSOR IN DCOFF" << std::endl;
+                new_closure.push_back(fcall._closure[i]);
+                
+            } else {
+                new_closure.push_back(m.weights_to_unpack.count(i) ? fcall._lazy_closure[i].eval() : fcall._closure[i]);
+                if (m.weights_to_unpack.count(i)) {
+                    fcall._lazy_closure[i].detach();
```
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/partitioning/patterns/dcoff.cpp:148]: Looks like those detaches in dcoff resolve memory consumption issue

File: src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp
Comment (smirnov-alexey) [inline src/plugins/intel_npu/src/plugin/npuw/lazy_tensor.cpp:1]: I just moved declaration to .hpp, so the only change in this file is `get_transformations()` method

**Review by esmirno — APPROVED** [review]

**Review by praasz — APPROVED** [review]
Ok, related to (reverted) core changes
